text/0046-metadata-for-offline-signers.md

```diff
@@ -0,0 +1,274 @@
+
+## Motivation
+
+### Background
+
+While all blockchain systems support (at least in some sense) offline signing used in air-gapped wallets and lightweight embedded devices, only few allow simultaneously complex upgradeable logic and full message decoding on the cold off-line signer side; Substrate is one of these heartening few, and therefore - we should build on this feature to greatly improve transaction security, and thus in general, network resilience.
+
+As a starting point, it is important to recognise that prudence and due care are naturally required. As we build further reliance on this feature we should be very careful to make sure it works correctly every time so as not to create false sense of security.
+
+In order to enable decoding that is small and optimized for chain storage transactions, a metadata entity is used, which is not at all small in itself (on the order of half-MB for most networks). This is a dynamic data chunk which completely describes chain interfaces and properties that could be made into a portable scale-encoded string for any given network version and passed along into an off-chain device to familiarize it with latest network updates. Of course, compromising this metadata anywhere in the path could result in differences between what user sees and signs, thus it is essential that we protect it.
+
+Therefore, we have 2 problems to be solved:
+
+1. Metadata is large, takes long time to be passed into a cold storage device with memory insufficient for its storage; metadata SHOULD be shortened and transmission SHOULD be optimized.
+2. Metadata authenticity SHOULD be ensured.
+
+As of now, there is no working solution for (1), as the whole metadata has to be passed to the device. On top of this, the solution for (2) heavily relies on a trusted party managing keys and ensuring metadata is indeed authentic: creating poorly decentralized points of potential failure.
+
+### Solution requirements
+
+#### Include metadata digest into signature
+
+Some cryptographically strong digest of metadata MAY be included into signable blob. There SHALL NOT be storage overhead for this blob, nor computational overhead, on the node side; thus MUST be a constant within given runtime version, deterministically defined by metadata.
+
+ - Metadata information that could be used in signable extrinsic decoding MAY be included in digest, its inclusion MUST be indicated in signed extensions;
+ - Digest MUST be deterministic with respect to metadata;
+ - Digest MUST be cryptographically strong against pre-image, both first and second;
+ - Extra-metadata information necessary for extrinsic decoding and constant within runtime version MUST be included in digest;
+ - It SHOULD be possible to quickly withdraw offline signing mechanism without access to cold signing devices;
+ - Digest format SHOULD be versioned.
+ - Work necessary for proving metadata authenticity MAY be omitted at discretion of signer device design (to support automation tools).
```

Comment:
I'd say that the device may choose not to verify the data. It'd be like current situation with PJS extension for example, that gets the metadata and uses it to interpret the extrinsic, without verifying its authenticity (probably because it can't).

Reply:
It's something we have no chance of enforcing here really. So we just state that it is not provable at this part of the architecture and expect the possibility of absence of this check when designing security strategies (for example, this would be required check in app code audits).

Also, tools like staking payout machine do not really need to check any metadata authenticity since they are limited to small number of calls and are often connected to petty cash accounts that are not worth being attacked anyway. We have an expetimental project ourselves where a signer resides on a physical key for physical lock, it does not need metadata verification as it only signs very narrow range of extrinsics (and it would severely degrade performance to do the check as hardware there is really humble); this is alternative - and very centralized in a sense, you just trust specialized firmware developer - approach to the same problem. To support signers like these we need to keep this optional.

Reply:
Just as a reminder to everybody that this is a fact - you should not rely on this actually happening. Explicit statement that proving that the check actually happened is not planned at all, just that the tool that could have done it had all required information in genuine form. I think this is important for future security analysts.